### PR TITLE
Clean up browser test APIs

### PR DESCRIPTION
**Pull request by Pessimistress**
#### Change List

- Removed a bunch of low-level public API from `BrowserDriver`; some server config keys are renamed to match node.js documentation (see updated docs)
- Remove need for `callExposedFunction` API (wait for `exposeFunction` promises to resolve before opening page)
- Consolidate node/browser communication to one function `window.browserTest`
- Now safe to import in browser - remove `module.require` hacks, use full es6 syntax
- Remove hack for port binding in `startServer`
- Automatically pipe console logs
- Better error handling for both the browser and server processes
- New `BrowserTestDriver` config flag `headless`. Non-headless mode behaves the same as the old browser tests. Headless mode does not launch a browser window, and exits the process on test completion even if tests failed.

### CODE REVIEW

File: docs/api-reference/test-utils/browser-driver.md [hunk 1]
```diff
@@ -6,115 +6,108 @@
...
-### startBrowser
+* `opts` (Object)
+  - `id` (String) - an id for this `BrowserDriver` instance. Default `browser-driver`.
 
-`driver.startBrowser(options)`
 
-Returns a `Promise` that resolves when the browser has started.
+## Methods
 
+### startBrowser(options)
```
Comment (ibgreen) [inline docs/api-reference/test-utils/browser-driver.md]: Just FYI:  I am using the convention of documenting types in the `###` header, and dropping the extra line of the options object in the argument list, in favor of `options.---`

E.g:

```
### startBrowser(options : Object)
`options.url` - 
```

I find it makes things a bit more compact, less duplicated and easier to scan.

I realize, we have a mixed style and no agreed-upon direction in this matter, so also fine as is.

File: docs/api-reference/test-utils/browser-driver.md [hunk 2]
```diff
@@ -6,115 +6,108 @@
...
+Parameters:
 
-* `functionToExpose`=`sendMessage` (`String`)
-Returns: a `Promise` that resolves when the browser code calls the function to the string the browser script passed to the exposed function.
+* `options` (Object)
+  - `url` (String) - The url to load in the page. Default `http://localhost`.
+  - `exposeFunctions` (Object) - keys are function names to be added to the page's `window` object, and the values are callback functions to execute in Node.js. See [exposeFunction](https://github.com/GoogleChrome/puppeteer/blob/v1.11.0/docs/api.md#pageexposefunctionname-puppeteerfunction) for details.
+  - `onLoad` (Function) - callback when the page is loaded
+  - `onConsole` (Function) - callback when the page logs to console
+  - `onError` (Function) - callback when the page crashes
```
Comment (ibgreen) [inline docs/api-reference/test-utils/browser-driver.md]: Nit: Sounds dramatic: "when the page crashes" ? => "called by puppeteer if the page crashes"?

File: docs/api-reference/test-utils/browser-driver.md [hunk 3]
```diff
@@ -6,115 +6,108 @@
...
-* `options`=`{maxBuffer: 5000 * 1024}`
-
-Example:
 ```js
 driver.startServer({
-  process: './node_modules/.bin/webpack-dev-server',
-  parameters: ['--config', 'test/render/webpack.config.js'],
-  options: {maxBuffer: 5000 * 1024}
+  command: './node_modules/.bin/webpack-dev-server',
+  arguments: ['--config', 'test/webpack.config.js'],
```
Comment (ibgreen) [inline docs/api-reference/test-utils/browser-driver.md]: Nit: I personally think we should move `test/webpack.config.js` to the root folder in our repos. That is unrelated to this text of course, though I see no reason to emphasize `test` subdir here?

File: docs/api-reference/test-utils/browser-driver.md [hunk 4]
```diff
@@ -6,115 +6,108 @@
...
-### stopServer
+Returns a `Promise` that resolves to the service URL when the server is available.
 
-Stopsthe server (kills the child process)
 
-`driver.stopServer()`
+### stopServer()
 
-Returns: nothing
+Stops the server (kills the child process).
```
Comment (ibgreen) [inline docs/api-reference/test-utils/browser-driver.md]: Nit: Does it actually kill the child process? It should but I think have had problems with this.

File: docs/api-reference/test-utils/browser-test-driver.md [hunk 5]
```diff
@@ -4,65 +4,97 @@
 
 # BrowserTestDriver (Test Automation Class)
```
Comment (ibgreen) [inline docs/api-reference/test-utils/browser-test-driver.md]: Unrelated nit: Ocular-gatsby auto extracts title from first markdown line, can we move the badge down?

File: docs/api-reference/test-utils/browser-test-driver.md [hunk 6]
```diff
@@ -4,65 +4,97 @@
...
 ```js
-// This is the script that runs in Node.js and starts the browser
-import {callExposedFunction} from 'probe.gl/test-utils';
-// Log to terminal console
-callExposedFunction('browserTestLog', 'Test started');
+// Run test cases
 ...
 // App is done running, terminate the browser instance
-callExposedFunction('browserTestComplete', {success: true});
+window.browserTest('done', 'All tests passed');
```
Comment (ibgreen) [inline docs/api-reference/test-utils/browser-test-driver.md]: Total nit: Not sure the new name is more descriptive/intuitive.

Maybe `window.browserTestDriver.sendMessage(...)`?

File: src/test-utils/browser-automation/browser-driver.js [hunk 7]
```diff
@@ -18,137 +18,180 @@
...
+      });
   }
 
-  exit() {
-    return Promise.resolve()
-      .then(() => this.stopBrowser())
-      .then(() => {
-        this.stopServer();
-        this.exitProcess();
+  _getAvailablePort() {
```
Comment (ibgreen) [inline src/test-utils/browser-automation/browser-driver.js]: I'd put this util in a separate file. Not sure how Mac specific this, maybe a fallback?

File: src/test-utils/browser-automation/browser-test-driver.js [hunk 8]
```diff
@@ -17,88 +17,146 @@
...
+
+    this.logger.log({
+      message: `${title}`,
+      color: COLOR.BRIGHT_YELLOW
+    })();
+
+    return this._startServer(config.server)
+      .then(url => {
+        this.logger.log({
+          message: `Started service at ${url}`,
```
Comment (ibgreen) [inline src/test-utils/browser-automation/browser-test-driver.js]: Nit: Can we say more about which service was started?

File: src/test-utils/browser-automation/browser-test-driver.js [hunk 9]
```diff
@@ -17,88 +17,146 @@
...
+
+  /* eslint-disable no-console */
+  _onConsole(event) {
+    if (!this.headless) {
+      // Do not mirror console messages if the browser is open
+      return;
+    }
+
+    // Crop very long text messages
+    const text = event.text().slice(0, 500);
```
Comment (ibgreen) [inline src/test-utils/browser-automation/browser-test-driver.js]: Nit: Use constant for magic threshold.

File: docs/api-reference/test-utils/browser-test-driver.md (hunk 6 above)
Comment (Pessimistress) [reply docs/api-reference/test-utils/browser-test-driver.md]: We can't expose objects, only functions to the global scope. How about `window.sendBrowserTestMessage`?

File: docs/api-reference/test-utils/browser-driver.md (hunk 4 above)
Comment (Pessimistress) [reply docs/api-reference/test-utils/browser-driver.md]: It does if you reach this method, or manually kill the browser test process (Ctrl+C). If the browser test script crashes then no. We need to listen to the process events and handle that.

File: docs/api-reference/test-utils/browser-test-driver.md (hunk 6 above)
Comment (ibgreen) [reply docs/api-reference/test-utils/browser-test-driver.md]: Sure let's go with that style.

Also I had a suggestion (I think in the other diff) regarding exposing multiple functions rather than generic sendMessage with overloaded args and return types. 

It seems to me the more generic `sendMessage` naming approach would be more appropriate if this system is designed to be user extensible on the other side, 

Maybe it is in terms of subclassing these drivers?

Comment (Pessimistress) [reply docs/api-reference/test-utils/browser-test-driver.md]: I agree, once we add another function for diffing it doesn't make sense to have a `sendMessage` func any more. I'm going to split this into two (`browserTestDriver_fail`, `browserTestDriver_finish`).
